### DISCUSSION TITLE
how can i filter sysex messages

### DISCUSSION THREAD

**Discussion by djnaoki-mrn** [tttapa/Control-Surface#554] [Q&A]
hi, i'm working on a midi/sysex project and i'm using this library to route the messages from my controller (behringer X-Touch in Xctl mode) to my daw (FL Studio).
in this mode, the X-touch send various sysex message to fl studio. 
i would like to know if it possible to filter thoes messages and catch them for delete them before they are routed to the usb ? i tried to use callback function (onSysExMessage), i can catch the messages, but i can't block them.
if you have a solution ? 
have a good day

**Comment by tttapa** [comment]
You can inherit from the `MIDI_Pipe` class and override the [`MIDI_Pipe::mapForwardMIDI(SysExMessage)`](https://tttapa.github.io/Control-Surface-doc/Doxygen/df/d72/classMIDI__Pipe.html#a2e64fcb4d7c5011fb59e63535568f968) method, filtering out the message you don't need.

Then connect the MIDI interfaces using such a custom pipe.

For example:
```cpp
class FilteringMIDI_Pipe : public MIDI_Pipe {
  void mapForwardMIDI(SysExMessage msg) override {
    if (/* not filtered out */)
      sourceMIDItoSink(msg);
  }
};
```

**Reply by djnaoki-mrn** [reply]
ho ! thank you so much ! it works like a charm ! i've spent a lot of time today for understand how it works, and it was pretty simple finally. thank you again :)
